contracts/adapters/FixedRebalanceExtension.sol

```diff
@@ -0,0 +1,483 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    SPDX-License-Identifier: Apache License, Version 2.0
+*/
+
+pragma solidity 0.6.10;
+pragma experimental ABIEncoderV2;
+
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import { BaseExtension } from "../lib/BaseExtension.sol";
+import { IBaseManager } from "../interfaces/IBaseManager.sol";
+import { ISetToken } from "../interfaces/ISetToken.sol";
+import { ICErc20 } from "../interfaces/ICErc20.sol";
+import { INotionalTradeModule } from "../interfaces/INotionalTradeModule.sol";
+import { INotionalProxy, MarketParameters } from "../interfaces/INotionalProxy.sol";
+import { IWrappedfCashComplete } from "../interfaces/IWrappedfCash.sol";
+import { IWrappedfCashFactory } from "../interfaces/IWrappedfCashFactory.sol";
+import { PreciseUnitMath } from "../lib/PreciseUnitMath.sol";
+
+
+
+
+/**
+ * @title FixedRebalanceExtension
+ * @author IndexCoop
+ *
+ * Smart contract that enables rebalancing of FIXED products
+ * Will sell redeem fCash positions and mint underweight fCash positions via NotionalTradeModule
+ */
+contract FixedRebalanceExtension is BaseExtension {
+    using PreciseUnitMath for uint256;
+    using SafeMath for uint256;
+
+
+    struct CallArguments {
+        address setToken;
+        uint16 currencyId;
+        uint40 maturity;
+        uint256 minMintAmount;
+        address sendToken;
+        uint256 sendAmount;
+    }
+
+    // /* ============ Events ============ */
+
+    event AllocationsUpdated(uint256[] _maturities, uint256[] _allocations);
+    event MaturityStatusSet(uint256 _maturity, bool _newStatus);
+
+    // /* ============ State Variables ============ */
+
+    mapping(uint256 => bool) internal isValidMaturity;       // Mapping of valid maturities
+    uint256[] validMaturities;                               // Array of valid maturities
+    ISetToken setToken;                                      // Instance of leverage token
+    INotionalTradeModule notionalTradeModule;                // Instance of leverage token
+    INotionalProxy notionalProxy;                            
+    IWrappedfCashFactory wrappedfCashFactory;                // Instance of leverage token
+    address underlyingToken;
+    address assetToken;
+    uint256[] maturities;                                    // Array of relative maturities in seconds (i.e. 3 months / 6 months)
+    uint256[] allocations;                                   // Relative allocations 
+    uint16 public currencyId;
+    bool tradeViaUnderlying;
```

Comment:
```suggestion
    // /* ============ State Variables ============ */

    mapping(uint256 => bool) internal isValidMaturity;           // Mapping of valid maturities
    uint256[] internal validMaturities;                          // Array of valid maturities
    uint256[] internal maturities;                               // Array of relative maturities in seconds (i.e. 3 months / 6 months)
    uint256[] internal allocations;                              // Relative allocations 

    ISetToken internal immutable setToken;                       // Instance of leverage token
    INotionalTradeModule internal immutable notionalTradeModule; // Instance of leverage token
    INotionalProxy internal immutable notionalProxy;                            
    IWrappedfCashFactory internal immutable wrappedfCashFactory; // Instance of leverage token
    address internal immutable underlyingToken;
    address internal immutable assetToken;
    uint16 public immutable currencyId;
    
    bool internal tradeViaUnderlying;
```

I think some of these could be better to have with public visibility, but this is the less controversial change.